### ISSUE TITLE
profileDir doesn't create profile

### ISSUE THREAD

**Issue by Spown** [hbenl/vscode-firefox-debug#18]
As was suggested in #10 the ``profileDir`` option can be used to store FF profile in the ``.vscode`` folder. However I only got to try it now and I think ``profileDir`` have some issues resolving paths... 
> "profileDir": "${workspaceRoot}/.vscode/vscode-firefox-debug-profile/"

This path doesn't exist so according to docs it should be created, instead I get:

![](https://dl.dropboxusercontent.com/u/105338698/ShareX/2016-11/Code_2016-11-11_21-52-35.png)

> "profileDir": "${workspaceRoot}/.vscode/vscode-firefox-debug-profile" //without trailing "/"
> "profileDir": "${workspaceRoot}\\\\.vscode\\\\vscode-firefox-debug-profile"

doesn't work either

If I create this folder manually, it starts to work, except the folder never gets populated and remains empty. I guess the debugger still uses temp profile.

**Comment by hbenl** [comment]
I recently changed the behaviour so that the debug adapter always makes a copy of the specified profile and works with that and didn't adjust the documentation accordingly.
Do you actually need to work with the specified profile directly instead of with a copy? The upside would be that you could make permanent changes to the Firefox configuration while debugging, otherwise you'd have to manually start Firefox with this profile directory in order to make permanent changes. If you need this then I could add an option to modify and work with the specified profile directly.

**Comment by Spown** [comment]
I... don't  quite follow... Copy of what and where? I'm talking about ``profileDir`` not ``profile``. I thought if I specify ``profileDir`` that doesn't exist - it would create it and it doesn't have to be a path under ``%APPDATA%\Mozilla\Firefox\Profiles`` - it can be everywhere, isn't it right?

**Comment by Spown** [comment]
Ok, yes, now I get it. You always make a copy (except when no profile is specified, I guess). Yes, I would say that this behavior is kinda counter intuitive. The user expects to use the specified ``profile`` directly and the specified ``profileDir`` directly. I would suggest using profiles directly by default and copy only if a specific option (let's say ``"proxyProfile": true||false``) is specified, like the following:

1. if no ``profile`` is specified: a clean new preconfigured profile would be created under temp and deleted afterwards. ``proxyProfile`` would be ignored.
2. if a non existent profile name is specified: it would be created in the profile collection, preconfigured and it would be used directly.
    1. If ``proxyProfile`` is set to true it would copy it into temp and work with the copy.
3. just as with 2. if ``profileDir`` is set to a path that contains no profile: it would create a preconfigured profile under that path and use it directly.
    1. If an existent ``profile`` name is specified: it would copy & preconfigure it to that path.
    2. If ``proxyProfile`` is set to true: the newly created profile under ``profileDir`` would get copied into temp and it becomes the working one.

So yeah, it can become redundant if the user would specify all 3 options at once, but it would make more sence from the get go. Like specifying ``profile`` and ``profileDir`` would get a copy of a profile into the desired location after the first lounch and by removing ``profile`` from the lounch config the user can work with that copy from now on and by switching ``proxyProfile`` he can kinda enter a safe mod.

Right now ``profileDir`` is a supplement option to ``profile`` but I suggest to make it work on it's own.

**Comment by hbenl** [comment]
I don't want to make the option of working with the specified profile directly the default, because I'm modifying the profile and can't undo the modifications (and some of those modifications are considered a potential security risk if you use them for regular web browsing).
By the way, `profileDir` is not a supplement to `profile` at all but an alternative. If you specify both options, `profileDir` will be used and `profile` will be ignored.

**Comment by Spown** [comment]
> I don't want to make the option of working with the specified profile directly
>Do you actually need to work with the specified profile directly instead of with a copy?

Ok, than something like ``useProfileDirectly`` would be nice... I guess it would start faster as well.

> By the way, profileDir is not a supplement to profile at all but an alternative.
> If you specify a profile directory which doesn't exist yet, it will be created and configured for remote debugging automatically.

How exactly? As I mentioned in the OP if I just specify some path (be it an existent or none) - I get an error or the default behavior (clean temp profile, no usage of the path).

**Comment by hbenl** [comment]
First of all you'd have to let Firefox create a profile directory by starting it with the command line option [`-profile`](https://developer.mozilla.org/en-US/docs/Mozilla/Command_Line_Options#-profile_.22.2Fpath.2Fto.2Fprofile.22). Then it should work.
If you create an empty profile directory (as you mentioned in the OP), that directory won't get filled because the debug adapter will make a copy of that empty directory and work with that - and that happens to be exactly the default behavior.

**Comment by hbenl** [comment]
(or you could manually copy one of your existing profile directories and work with that)

**Comment by Spown** [comment]
> If you create an empty profile directory (as you mentioned in the OP), that directory won't get filled because the debug adapter will make a copy of that empty directory and work with that - and that happens to be exactly the default behavior.

No, it fails to do anything. At first I simply specified a path with ``profileDir``. Without creating an empty dir manually. it resulted in an error. only then I tried to create the path manually. As it states in your docs:  "If you specify a __profile directory__ which doesn't exist yet, it will be created and configured for remote debugging automatically." Does it this piece of texts describe ``profile`` or ``profileDir`` functionality? Or both? Because if you say "profile directory" I (the user) think of ``profileDir``.

> First of all you'd have to let Firefox create a profile directory by starting it with the command line option -profile. Then it should work.

Ok, now I'm really confused. You said ``profileDir`` isn't supplementary to ``profile`` but now you say I have to manually create a profile in order to use ``profileDir``. So does this mean I have to define this (previously, manually created) profile's name under ``profile``? Or how exactly does ``profileDir`` know which profile to copy? Why does ``profileDir`` even needs an existent profile? Again: "If you specify a profile directory __which doesn't exist__ yet, it will be __created__ and configured for remote debugging __automatically__."

**Comment by hbenl** [comment]
> As it states in your docs: "If you specify a profile directory which doesn't exist yet, it will be created and configured for remote debugging automatically." Does it this piece of texts describe profile or profileDir functionality? Or both? Because if you say "profile directory" I (the user) think of profileDir.

Yes, this did describe `profileDir`. But that's where I changed the behaviour, the `profileDir` doesn't create a profile directory at the given path anymore. Your quote from the documentation was outdated and I have removed it now. Sorry about the confusion.

Both `profile` and `profileDir` reference *existing* profile directories: `profile` references a profile directory created using Firefox' profile manager under `%APPDATA%\Mozilla\Firefox\Profiles`. `profileDir` can reference a profile directory anywhere in the filesystem.
These two settings correspond to Firefox' command line options: `profile` corresponds to [`-P`](https://developer.mozilla.org/en-US/docs/Mozilla/Command_Line_Options#-P_.22profile_name.22) and `profileDir` corresponds to [`-profile`](https://developer.mozilla.org/en-US/docs/Mozilla/Command_Line_Options#-profile_.22.2Fpath.2Fto.2Fprofile.22). But whereas these command line options use the profile directories directly (and create them if they don't exist), the debug adapter always makes a temporary copy of them (and fails if they don't exist).

I hope this cleared things up (?)
